### DISCUSSION TITLE
Information: Changing URL without reloading

### DISCUSSION THREAD

**Discussion by GDuroS** [anvil-works/routing#52] [Q&A] [answered]
In anvil_extras.routing, when you navigate passing redirect=True, the url is changed without reloading any forms.

Reading the docs I saw that everytime any query or path is changed, the form is reloaded. Is there a way of changing the **path** without loading a new instance of the form? Or only when changing **query params** AND reimplementing cache_deps?

One thing I didn't find in the docs: when you navigate using only `navigate(hash="something")`, does this also reloads the form? If it does, is it affected by cache_deps?

Also, can you clarify how the cache_deps works? In the docs, the only example returns None. Can you give an example of how to return values in cache_deps and what would that achieve?

Thanks in advance.

**Comment by s-cork** [comment, marked as answer]
There is no way to change the path without navigating, what's the use case?

---

query parameters might be used to encapsulate the state of the page

```text
my_app.anvil.app/users/user?tab=info&page=1&show_sidebar=true
```

let's say you had a route defined like this

```python
Route.cache_form = True

class UserRoute(Rote):
    path = "/users/:id"
    pages = "Pages.User"
    
    def cache_deps(self, query, **loader_args):
        return {"tab": query.get("tab", "default")}

```

cache deps determine how to cache forms
in this example if tab could be `"default"`, `"info"` or `"activity"`
then the cache system would cache based on the path and the value of the tab here.
i.e. we would get 3 different form instances, 1 for each of `user/123`, `user/123?tab=info` and `user/123?tab=activity`
but things like `page` and `show_sidebar` should be ignored when considering caching
(A form instance can listen to changes to query parameters and change its state accordingly)

If i set it to `None` then i'm saying that query parameters do not contribute to how forms are cached, only the path.

---

Honestly I haven't really thought much about the hash parameter, I think it would be rare to need it
Generally the hash parameter is used to ensure a certain element with a certain id is scrolled into view
I would use query parameters for page state
A hash change should cause a navigation, but since cache_deps don't do anything with the hash by default
it'll load a cached form if it exists

**Reply by GDuroS** [reply]
Originally, I had a page that works both as just vieweing and editing a registry, updating with binds to change modes seemlessly to the user. The paths were: `#registry/edit/:id` and `#registry/view/:id`. This worked with hashrouting because obviusly with was just hash, everything is just the same page. While I was asking I already was changing the structure to user queries now, so it would be `/registry/:id` and `/registry/:id?mode=edit`. It's working perfectly.

---
So, in summary, I use `parse_query` to "remove" any invalid queries and to return the valid queries **used** in the navigation, and in cache_deps I return **all** the possible query values and how should I get then (and default value, but basically they will just be the tag and get the tag from query).

So if my `parse_query` method allows queries for "tabs" with values `"1"`, `"2"` and `"3"`and "mode" with values `"edit"` and `"view"`, but my `cache_des` method only returns `{"mode": query.get("mode", "view")}` that means that the page will load and select tab accordinly using the tabs query, but the cache and not reloading will happen only when changing modes in this way. Is that it?

If that's it, I think I understood. This other example could go to the docs, I believe. It's always good to see an example using a valid return value (other than `None`) to understand more how the code works.

---
In my humble opinion, I think hash should be used how you said it, only for in-page navigation like scrolling to a specific section of the same page. That way, I think it should **never** reload the page when changing. It souldn't be used for navigating out of the current form, only for scrolling to a specific component. Maybe even a method like `mark_component_for_hash(component:Component, hash:str)` so when raising the `hash_changed` event the default behaviour would be to call `component.scroll_into_view()`. But this could be just left to the users discretion.

I also don't think I will use often. I rarelly create longer forms that would benefit from it - users usually don't like to scroll to much - but maybe I can see use for it if it's available.

---
In any case, thank you for you answer (and sorry for posting as an issue before, I'm not really that familiar with GitHub since until recently I just coded for my employer (which uses GitLab) and never did PRs and things like it. Thank for understanding too.

**Reply by s-cork** [reply]
Yeah that all sounds about right. Note that if you've already parsed the query yourself then the query argument in `cache_deps` will be the parsed query, which might simplify your code slightly. 

So yes if your cache_deps method  returns a dict with mode as the only key then you will have 2 cached versions of your form. One where the mode is edit and one where the mode is view. 

---

Yes we should add more examples in the docs. PRs are always welcome 😉 

---

Yeah, as and when hash issues/features get requested we'll start adding things. I do think that most use cases will be content with just path and query.

**Reply by GDuroS** [reply]
Sure. I can take this as a PR and add a example in the doc. Just gonna fork and ready a few other things and make a PR.
